layer.py

```diff
@@ -417,7 +421,105 @@ def copy(self, target_path, copied_files, keep_existent=False):
             self._change_data_source(new_source)
         return copied_files
 
-    def _change_data_source(self, new_data_source):
+    def convert_to_gpkg(self, target_path):
+        """
+        Convert a layer to geopackage in the target path and adjust its datasource. If
+        a layer is already a geopackage, the dataset will merely be copied to the target
+        path.
+
+        :param layer: The layer to copy
+        :param target_path: A path to a folder into which the data will be copied
+        :param keep_existent: if True and target file already exists, keep it as it is
+        """
+
+        if not self.layer.type() == QgsMapLayer.VectorLayer:
+            return
+
+        parts = None
+        file_path = ""
+        suffix = ""
+        uri_parts = self.layer.source().split("|", 1)
+        if len(uri_parts) > 1:
+            suffix = uri_parts[1]
+
+        if self.layer.dataProvider() is not None:
+            metadata = QgsProviderRegistry.instance().providerMetadata(
+                self.layer.dataProvider().name()
+            )
+            if metadata is not None:
+                parts = metadata.decodeUri(self.layer.source())
+                if "path" in parts:
+                    file_path = parts["path"]
+        if file_path == "":
+            file_path = uri_parts[0]
+
+        dest_file = ""
+        new_source = ""
+        if os.path.isfile(file_path):
+            # check if the source is a geopackage, and merely copy if it's the case
+            source_path, file_name = os.path.split(file_path)
+            if file_name.endswith(".gpkg"):
```

Comment:
Any chance to use something like:
```
QgsVectorDataProvider::storageType()
```


<br class="Apple-interchange-newline">

Reply:
@suricactus , completely forgot about this one, good call.